shell/common/asar/archive_win.cc

```diff
@@ -0,0 +1,179 @@
+// Copyright 2023 Slack Technologies, Inc.
+// Contributors: Weiyun Dai (https://github.com/WeiyunD/), Andrew Lay
+// (https://github.com/guohaolay) Use of this source code is governed by the MIT
+// license that can be found in the LICENSE file.
+
+#include "shell/common/asar/archive.h"
+
+#include <algorithm>
+#include <sstream>
+
+#include "base/json/json_reader.h"
+#include "base/logging.h"
+#include "base/no_destructor.h"
+#include "base/strings/string_util.h"
+#include "base/strings/string_util_win.h"
+#include "base/strings/utf_string_conversions.h"
+#include "shell/common/asar/asar_util.h"
+
+using namespace std;
+using namespace base;
+
+#define INTEGRITY_CHECK_RESOURCE_TYPE L"Integrity"
+#define INTEGRITY_CHECK_RESOURCE_ITEM L"ElectronAsar"
+
+namespace asar {
+
+/*
+  Windows integrity payload cache
+    - key: file rel path
+    - value: integrity payload
+  absl::nullopt: cache not initialized
+  empty map: cache initialized but no data.
+*/
+static base::NoDestructor<
+    absl::optional<unordered_map<string, IntegrityPayload>>>
+    s_win_integrity_payloads_cache;
+
+absl::optional<base::FilePath> Archive::RelativePath() const {
+  // In case user has Windows long path enabled, the exe path should still no
+  // larger than the asar path.
+  size_t buf_len = max(path_.value().size(), (size_t)MAX_PATH);
+  wchar_t* buf = new (std::nothrow) wchar_t[buf_len];
+  if (!buf) {
+    LOG(FATAL) << "Failed to alloc enough space to store exe file path";
+    return absl::nullopt;
+  }
+
+  DWORD len = GetModuleFileName(NULL, buf, buf_len);
+  wstring exe_path = wstring(buf, len);
+  delete[] buf;
+
+  if (!len) {
+    LOG(FATAL) << "GetModuleFileName failed. Last error: " << GetLastError();
+    return absl::nullopt;
+  }
+
+  base::FilePath exe_dir = base::FilePath(exe_path).DirName();
+  base::FilePath relative_path;
+  if (!exe_dir.AppendRelativePath(path_, &relative_path)) {
+    return absl::nullopt;
+  }
+
+  return relative_path;
+}
+
+void LoadIntegrityConfigCache() {
+  // Skip loading if cache is already loaded
+  if (s_win_integrity_payloads_cache->has_value()) {
+    return;
+  }
+
+  // Init cache
+  *s_win_integrity_payloads_cache = unordered_map<string, IntegrityPayload>();
+
+  // Load integrity config from exe resource
+  HMODULE module_handle = ::GetModuleHandle(NULL);
+
+  HRSRC resource = ::FindResource(module_handle, INTEGRITY_CHECK_RESOURCE_ITEM,
+                                  INTEGRITY_CHECK_RESOURCE_TYPE);
+  if (!resource) {
+    LOG(FATAL) << "FindResource failed. Last error: " << GetLastError();
```

Comment:
Chromium has `PLOG` for automatically printing out the error:
```suggestion
    PLOG(FATAL) << "FindResource failed";
```

Reply:
Updated